File: apps/api/plane/app/serializers/project_template.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Third party imports
from rest_framework import serializers

# Module imports
from plane.app.serializers.base import BaseSerializer
from plane.db.models import Issue, Module, WorkspaceProjectTemplate


class WorkspaceProjectTemplateSerializer(BaseSerializer):
    class Meta:
        model = WorkspaceProjectTemplate
        fields = [
            "id",
            "name",
            "description",
            "seed_id",
            "workspace_type",
            "is_system",
            "is_default_for_type",
            "project_id",
            "sort_order",
            "logo_props",
            "archived_at",
            "is_hidden",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields

    def to_representation(self, instance):
        data = super().to_representation(instance)
        data["project_id"] = str(instance.project_id)
        data["module_count"] = getattr(
            instance,
            "module_count",
            Module.objects.filter(project=instance.project, deleted_at__isnull=True).count(),
        )
        data["issue_count"] = getattr(
            instance,
            "issue_count",
            Issue.objects.filter(project=instance.project, deleted_at__isnull=True).count(),
        )
        return data


class WorkspaceProjectTemplateWriteSerializer(BaseSerializer):
    class Meta:
        model = WorkspaceProjectTemplate
        fields = ["name", "description", "logo_props", "sort_order", "is_hidden"]


class ProjectTemplateInstantiateSerializer(serializers.Serializer):
    name = serializers.CharField(max_length=255)
    identifier = serializers.CharField(max_length=12)
    description = serializers.CharField(required=False, allow_blank=True, default="")


class ProjectTemplatePreviewSerializer(serializers.Serializer):
    id = serializers.UUIDField()
    name = serializers.CharField()
    description = serializers.CharField()
    modules = serializers.ListField()
    states = serializers.ListField()
    issue_count = serializers.IntegerField()
    is_system = serializers.BooleanField()

File: packages/types/src/project-template.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

import type { EWorkspaceType } from "./workspace-type";

export type TProjectTemplateSeedRef = {
  seedId: string;
  wave: 1 | 2;
  isDefault?: boolean;
};

export type TProjectTemplateSeedMeta = {
  seedId: string;
  i18n: {
    en: string;
    "pt-BR": string;
  };
  i18n_description?: {
    en: string;
    "pt-BR": string;
  };
  workspaceTypes: EWorkspaceType[];
  wave: 1 | 2;
  moduleCount: number;
  issueCount: number;
  defaultIssueState: string;
  hasBacklog: boolean;
};

export type TProjectTemplateSeedState = {
  name: string;
  group: string;
  color?: string;
};

export type TProjectTemplateSeedLabel = {
  name: string;
  color?: string;
};

export type TProjectTemplateSeedModule = {
  name: string;
  sort_order?: number;
};

export type TProjectTemplateSeedIssue = {
  name: string;
  module?: string;
  labels?: string[];
  priority?: string;
  description?: string;
  parent?: string;
};

export type TProjectTemplateSeedDefinition = {
  seedId: string;
  workspaceTypes: string[];
  defaultIssueState: string;
  states: TProjectTemplateSeedState[];
  labels: TProjectTemplateSeedLabel[];
  modules: TProjectTemplateSeedModule[];
  issues: TProjectTemplateSeedIssue[];
};

export type IWorkspaceProjectTemplate = {
  id: string;
  name: string;
  description: string;
  seed_id: string | null;
  workspace_type: string | null;
  is_system: boolean;
  is_default_for_type: boolean;
  project_id: string;
  sort_order: number;
  logo_props: Record<string, unknown>;
  archived_at: string | null;
  is_hidden: boolean;
  module_count?: number;
  issue_count?: number;
  created_at: string;
  updated_at: string;
};

export type TWorkspaceProjectTemplatePayload = {
  name: string;
  description?: string;
  logo_props?: Record<string, unknown>;
  is_hidden?: boolean;
};

export type TProjectTemplateInstantiatePayload = {
  name: string;
  identifier: string;
  description?: string;
  project_lead?: string;
};

export type TProjectTemplatePreview = {
  id: string;
  name: string;
  description: string;
  modules: Array<{ name: string; issue_count: number }>;
  states: string[];
  issue_count: number;
  is_system: boolean;
};
